### PR TITLE
add tree-shakable providers doc

### PR DESCRIPTION
**Pull request by sreevani-ship-it**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[x] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## Other information
https://github.com/angular/angular/pull/22803 - Orginal PR

### CODE REVIEW

**Comment by about-code** [comment]
Current drafts on *treeshakable providers* in https://pr22803-95a505f.ngbuilds.io/guide/dependency-injection#creating-tree-shakable-providers suggest that a replacement of an injectable type should/could be declared in the same module which defines the injectable type, e.g. 
```
@Injectable({ 
  providedIn: 'root', 
  useClass: FooServiceImpl  // same with useFactory
}) 
export class FooService {}
```

What I've learnt about DI, though, is that DI is meant to decouple a service consumer (which depends on an interface type like `FooService`) from a concrete implementation like `FooServiceImpl` in order to be able to inject alternative implementations or mocks of the interface type into the consumer without having to alter service consumer nor service interface type (because I may neither be consumer nor interface type provider but a third-party providing an alternative implementation)

But if I do specify an implementation of an interface in the same module which declares the interface type then this does seem to me like a tight coupling of an interface type and a concrete implementation. What's the point of DI then, anymore? The drafts look like that by choosing and injecting the interface type one does also inevitably choose a concrete implementation. Then I could do `new FooServiceImpl()` in the service consumer, once again like I did in the "good" old days without DI.

Or should `FooServiceImpl` be considered more kind of a *default implementation* which could be overriden with an alternative implementation provided via some AppModule provider? But if this syntax is about default implementations is it required at all? I mean `FooService` could be that default implementation itself, then.

What am I missing? I do understand tree shaking and I actually do understand dependency injection but it seems to me like the concepts are fundamentally incompatible if something like the above is the only way to make it *somehow* work together. Would love to find out where I miss the point.

**Comment by jenniferfell** [comment]
@alxhub Can you provide some insight on the question from @about-code ? Thanks!

**Comment by alxhub** [comment]
@about-code you're not missing anything, you actually answered your own question.

> Or should FooServiceImpl be considered more kind of a default implementation which could be overriden with an alternative implementation provided via some AppModule provider?

Yes, exactly.

> But if this syntax is about default implementations is it required at all? I mean FooService could be that default implementation itself, then.

Yes, it could:

```
@Injectable({providedIn: 'root'})
export class FooService {
  ...
}
```

Every token must have a provider in order to be injected. `@NgModule` allowed you to define that provider in 4 ways: class, existing, factory, and value. `@Injectable` has the same level of flexibility, but moving the definition onto the type itself (as a sort of "default provider" as you called it) allows the type to be removed if it's not referenced.

**Comment by jenniferfell** [comment]
@alxhub : @chembu has incorporated the code snippets from your example. Time for approval review, meaning:
1) Make sure all changes you requested previously are incorporated
2) Check that any reviewer comments on this thread have been incorporated into doc to your satisfaction
3) Check for any show-stoppers. 
4) Indicate your approval so pull-approve passes. 

I'm doing the same as co-approver from doc team. 

Please review by EOD 3/30. We need to have any final critical changes incorporated by @chembu , your approval, and mark for merge by EOD 4/2. Thanks! (target is next RC)

**Comment by jenniferfell** [comment]
@IgorMinar Can you also provide an approval. We need an approval from you or someone else in the guide-and-tutorial group. This PR doesn't touch tutorial files, just ngmodule and DI files, but pull-approve is asking for the guide-and-tutorial group. Thanks!

File: aio/content/guide/dependency-injection.md [hunk 1]
```diff
@@ -572,6 +401,185 @@ Here you see the new and the old implementation side-by-side:
...
+
+In this example, to provide services in Angular, you include them in an `@NgModule`:
+
+<code-example path="dependency-injection/src/app/tree-shaking/service-and-module.ts"  title="src/app/tree-shaking/service-and-modules.ts" linenums="false"> </code-example>
+
+This module can then be imported into your application module, to make the service available for injection in your app:
+
+<code-example path="dependency-injection/src/app/tree-shaking/app.module.ts"  title="src/app/tree-shaking/app.modules.ts" linenums="false"> </code-example>
+
+When `ngc` runs, it compiles AppModule into a module factory, which contains definitions for all the providers declared in all the modules it includes. At runtime, this factory will become an Injector which will instantiate these services.
```
Comment (jenniferfell) [inline aio/content/guide/dependency-injection.md]: Edit: "At runtime, this factory becomes an injector that will instantiate these services."

File: aio/content/guide/dependency-injection.md [hunk 2]
```diff
@@ -572,6 +401,185 @@ Here you see the new and the old implementation side-by-side:
...
+
+The reason that tree-shaking breaks in above method is because Angular cannot decide to exclude one chunk of code (the provider definition for the service within the module factory) based on whether another chunk of code (the service class) is used. To make services tree-shakeable, the information about how to construct an instance of the service (the provider definition) needs to be a part of the service class itself.
```
Comment (jenniferfell) [inline aio/content/guide/dependency-injection.md]: Edit: "Tree shaking doesn't work in the method above because Angular..."

File: aio/content/guide/dependency-injection.md [hunk 3]
```diff
@@ -572,6 +401,185 @@ Here you see the new and the old implementation side-by-side:
...
+
+#### Creating tree-shakable providers
+
+To create providers that are tree-shakable, the information that used to be specified in the module should be specified in the `@Injectable` decorator on the service itself.
+
+Here is the tree-shakeable equivalent to the Service / ServiceModule example above:
```
Comment (jenniferfell) [inline aio/content/guide/dependency-injection.md]: Edit: "The following example shows the ..." and remove the spaces around the / or the / entirely and replace it with a word like "and"

File: aio/content/guide/dependency-injection.md [hunk 4]
```diff
@@ -572,6 +401,185 @@ Here you see the new and the old implementation side-by-side:
...
+
+<code-example path="dependency-injection/src/app/tree-shaking/service.ts"  title="src/app/tree-shaking/service.ts" linenums="false"> </code-example>
+
+Here, `providedIn` allows you to declare the injector which injects this service. Unless there is a special case the value should always be root. This ensures that the service will be scoped to the root injector, without needing to name a particular module that will be present in that injector.
```
Comment (jenniferfell) [inline aio/content/guide/dependency-injection.md]: Edit recommendations: "In the example above, providedIn allows..." "Unless there is a special case, the value..." "Setting the value to root ensures that the service is scoped to the root injector, without naming a particular module that is present in that injector."

File: aio/content/guide/dependency-injection.md [hunk 5]
```diff
@@ -572,6 +401,185 @@ Here you see the new and the old implementation side-by-side:
...
+  <code-pane title="hero-list.component (with DI)" path="dependency-injection/src/app/heroes/hero-list.component.2.ts">
+  </code-pane>
+
+  <code-pane title="hero-list.component (without DI)" path="dependency-injection/src/app/heroes/hero-list.component.1.ts">
+  </code-pane>
+</code-tabs>
+
+Notice that the `HeroListComponent` doesn't know where the `HeroService` comes from.
+_You_ know that it comes from the parent `HeroesComponent`.
+But if you decided instead to provide the `HeroService` in the `AppModule`,
```
Comment (jenniferfell) [inline aio/content/guide/dependency-injection.md]: Edit: Remove "But"

File: aio/content/guide/dependency-injection.md [hunk 6]
```diff
@@ -572,6 +401,185 @@ Here you see the new and the old implementation side-by-side:
...
+The _only thing that matters_ is that the `HeroService` is provided in some parent injector.
+
+{@a singleton-services}
+
+## Singleton services
+
+Services are singletons _within the scope of an injector_.
+There is at most one instance of a service in a given injector.
+
+There is only one root injector and the `UserService` is registered with that injector.
```
Comment (jenniferfell) [inline aio/content/guide/dependency-injection.md]: Edit: "There is only one root injector, and the ..."

File: aio/content/guide/dependency-injection.md [hunk 7]
```diff
@@ -572,6 +401,185 @@ Here you see the new and the old implementation side-by-side:
...
+Therefore, there can be just one `UserService` instance in the entire app
```
Comment (jenniferfell) [inline aio/content/guide/dependency-injection.md]: Edit: "...in the entire app, and..."

File: aio/content/guide/dependency-injection.md [hunk 8]
```diff
@@ -572,6 +401,185 @@ Here you see the new and the old implementation side-by-side:
...
+However, Angular DI is a 
+[hierarchical injection system](guide/hierarchical-dependency-injection), 
+which means that nested injectors can create their own service instances.
+Angular creates nested injectors all the time.
+
+{@a component-child-injectors}
+
+## Component child injectors
+
+For example, when Angular creates a new instance of a component that has `@Component.providers`,
```
Comment (jenniferfell) [inline aio/content/guide/dependency-injection.md]: Edit: Rearrange so there isn't a heading followed by "For example." This structure breaks up the context of what is being demonstrated. It also is confusing for someone who uses the right nav to jump to this section.

File: aio/content/guide/dependency-injection.md [hunk 9]
```diff
@@ -572,6 +401,185 @@ Here you see the new and the old implementation side-by-side:
...
+it also creates a new _child injector_ for that instance.
+
+Component injectors are independent of each other and
+each of them creates its own instances of the component-provided services.
+
+When Angular destroys one of these component instance, it also destroys the
```
Comment (jenniferfell) [inline aio/content/guide/dependency-injection.md]: Typo: Missing s in "one of these component instances"

File: aio/content/guide/dependency-injection.md [hunk 10]
```diff
@@ -572,6 +401,185 @@ Here you see the new and the old implementation side-by-side:
...
+component's injector and that injector's service instances. 
+
+Thanks to [injector inheritance](guide/hierarchical-dependency-injection),
```
Comment (jenniferfell) [inline aio/content/guide/dependency-injection.md]: Edit: Because of

File: aio/content/guide/dependency-injection.md [hunk 11]
```diff
@@ -649,52 +658,45 @@ One solution to choosing a provider token for non-class dependencies is
...
+<code-example>
+import { InjectionToken } from '@angular/core';
+export const TOKEN = new InjectionToken('desc');
 </code-example>
 
-The type parameter, while optional, conveys the dependency's type to developers and tooling.
-The token description is another developer aid.
-
-Register the dependency provider using the `InjectionToken` object:
+Like with `@Injectable`, you can directly configure a provider when creating an InjectionToken, that will determine which injector provides the token and how the value will be created. But, unlike with `@Injectable`, you cannot define standard providers (such as `useClass` or `useFactory`) with `InjectionToken`. Instead, you specify a factory function which returns the value to be provided directly:
```
Comment (jenniferfell) [inline aio/content/guide/dependency-injection.md]: Edit: I'm not sure of the right answer technically, but the paragraph above doesn't parse well and has grammar issues. One attempt: 

You can directly configure a provider when creating an InjectionToken. Directly configuring a provider determines which injector provides the token and how the value is created. Using an InjectionToken is similar to using @Injectable, expect that you cannot define standard providers such as useClass or useFactory. Instead, you specify a factory function that returns the value to be provided directly.

File: aio/content/guide/dependency-injection.md [hunk 12]
```diff
@@ -649,52 +658,45 @@ One solution to choosing a provider token for non-class dependencies is
...
 
-<code-example path="dependency-injection/src/app/providers.component.ts" region="providers-9"  linenums="false">
+<code-example>
```
Comment (jenniferfell) [inline aio/content/guide/dependency-injection.md]: Examples: Will these code snippets be coming from live examples?

**Review by jenniferfell — COMMENTED** [review]
There are some code snippets directly embedded here. Will those be coming from a live example soon?

Lots of notes, but I don't think there are any show stoppers. Some suggestions for clarifying sections that were hard to parse. Some replacements of which with that for accuracy. A few typos. If they don't make it in for this merge, please capture externally somehow for the next rev.

**Comment by about-code** [comment]
@alxhub Thank you for your clarification.

File: aio/content/guide/providers.md [hunk 13]
```diff
@@ -2,34 +2,47 @@
...
+
+The service itself is just a class the CLI has generated, that's decorated with `@Injectable`. By default, this decorator is configured with a `providedIn` property, which creates a provider for the service. In this case, `providedIn: 'root'` specifies that the service should be provided in the root injector.
 
 
 ## Provider scope
 
-When you add a service provider to the `providers` array of the root module, it’s available throughout the app. Additionally, when you import a module that has providers, those providers are also available to all the classes in the app as long they have the lookup token. For example, if you import the `HttpClientModule` into your `AppModule`, its providers are then available to the entire app and you can make HTTP requests from anywhere in your app.
+When you add a service provider to the root application injector, it’s available throughout the app. Additionally, these providers are also available to all the classes in the app as long they have the lookup token. 
+
+Unless there is a special case, you should always provide your service in the root injector.
```
[the remark this answers is not in the document]
Comment (alxhub) [reply aio/content/guide/providers.md]: There are two:

1) The service should only be available in some context which is not the root injector (e.g. it might make sense to scope an `AdminService` to the lazy loaded `Admin` module injector.

2) The author really wants the service to only be available if the consumer imports a particular `@NgModule`.

File: aio/content/guide/dependency-injection.md [hunk 14]
```diff
@@ -81,18 +81,17 @@ _injecting the service_ into the `HeroList` component.
...
 
 An Angular injector is responsible for creating service instances and injecting them into classes like the `HeroListComponent`.
 
 You rarely create an Angular injector yourself.
 Angular creates injectors for you as it executes the app,
 starting with the _root injector_ that it creates during the [bootstrap process](guide/bootstrapping).
 
-You do have to register _providers_ with an injector 
-before the injector can create that service.
+Angular doesn't automatically know how you want to create instances of your services, or in which injector. You must configure it by specifying providers for every service.
```
[the remark this answers is not in the document]
Comment (alxhub) [reply aio/content/guide/dependency-injection.md]: A service can be _provided_ more than once. I wouldn't say "the service has more than one provider", since those extra providers are declared independently of the service. The service can declare zero or one providers of its own (via `@Injectable`)

File: aio/content/guide/dependency-injection.md [hunk 15]
```diff
@@ -141,21 +145,31 @@ into any class that it creates.
...
+</code-example>
 
 {@a ngmodule-vs-comp}
 
-### _@NgModule_ or _@Component_?
+### @Injectable,, _@NgModule_ or _@Component_?
 
-Should you register a service with an Angular module or with a component?
-The two choices lead to differences in service _scope_ and service _lifetime_.
+Should you register a service with an `@Injectable` decorator, Angular module or with a component?
```
[the remark this answers is not in the document]
Comment (alxhub) [reply aio/content/guide/dependency-injection.md]: > Should you provide your service using the `@Injectable` decorator, in an `@NgModule`, or in a `@Component`?

**Review by alxhub — APPROVED** [review]
LGTM, commented a couple of times for technical clarity.

File: aio/content/guide/styleguide.md [hunk 16]
```diff
@@ -4685,8 +4685,7 @@ Compare with the less preferred `host` metadata alternative.
 
 
 
-**Why?** When providing the service to a top level component,
-that instance is shared and available to all child components of that top level component.
+**Why?** When providing the service to a root injector, that instance of the service is shared and available in every class that needs the service. Additionally, when you register a service in the `@Injectable` decorator of the service, optimization tools such as those used by the CLI's production builds can perform tree shaking and remove services that aren't used by your app.
```
[the remark this answers is not in the document]
Comment (chembu) [reply aio/content/guide/styleguide.md]: Done

File: aio/content/guide/ngmodule-faq.md [hunk 17]
```diff
@@ -198,7 +198,7 @@ Its only purpose is to add http service providers to the application as a whole.
 
 ## What is the `forRoot()` method?
 
-The `forRoot()` static method is a convention that makes it easy for developers to configure the module's providers.
+The `forRoot()` static method is a convention that makes it easy for developers to configure services and providers for a module which are intended to be singletons. For services it is preferable to specify `providedIn: 'root'` on the service's `@Injectable()` decorator, which has the same effect.
```
[the remark this answers is not in the document]
Comment (chembu) [reply aio/content/guide/ngmodule-faq.md]: Done

[the remark this answers is not in the document]
Comment (chembu) [reply aio/content/guide/ngmodule-faq.md]: Explained why to use  @Injectable() decorator

File: aio/content/guide/ngmodule-faq.md [hunk 18]
```diff
@@ -374,8 +376,9 @@ This means that lazy-loaded modules can't reach them.
 
 ## Should I add other providers to a module or a component?
 
-In general, prefer registering feature-specific providers in modules (`@NgModule.providers`)
-to registering in components (`@Component.providers`).
+Preferably, providers should be configured using `@Injectable` syntax. If possible, they should be provided in the application root (`providedIn: 'root'`). Services configured this way may still end up being lazily loaded if they are only used from a lazily loaded context.
```
[the remark this answers is not in the document]
Comment (chembu) [reply aio/content/guide/ngmodule-faq.md]: Other parts of DI doc explains why sometimes it is not possible to have an injector at the root level.

File: aio/content/guide/providers.md [hunk 19]
```diff
@@ -2,34 +2,47 @@
...
 
-## Create a service
-You can provide services to your app by using the `providers` array in an NgModule.
-Consider the default app generated by the CLI. In order to add a user service to it,
-you can generate one by entering the following command in the terminal window:
+A provider is an instruction to the DI system on how to obtain a value for a dependency. Most of the time, these dependencies will be services that you create and provide.
+
+## Providing a Service
+
+If you already have a CLI generated app, create a service using the following CLI command in the root project directory. Replace User with the name of your service.
```
[the remark this answers is not in the document]
Comment (chembu) [reply aio/content/guide/providers.md]: used underscore to differentiate User

File: aio/content/guide/providers.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (chembu) [reply aio/content/guide/providers.md]: Updated to include the special case

File: aio/content/guide/dependency-injection.md (hunk 11 above)
Comment (chembu) [reply aio/content/guide/dependency-injection.md]: Re-phrased.

**Comment by jenniferfell** [comment]
@IgorMinar Can you also provide an approval. We need an approval from you or someone else in the guide-and-tutorial group. This PR doesn't touch tutorial files, just ngmodule and DI files, but pull-approve is asking for the guide-and-tutorial group. AlexR has approved. Thanks!

**Review by IgorMinar — APPROVED** [review]

**Comment by IgorMinar** [comment]
lgtm
